components/script/dom/documentorshadowroot.rs

```diff
@@ -345,4 +358,112 @@ impl DocumentOrShadowRoot {
+    ///
+    /// The main purpose from this function is to set the `&mut adopted_stylesheet` to match
+    /// `incoming_stylesheet` and update the corresponding Styleset in a Document or a ShadowRoot.
+    /// In case of duplicates, the setter will respect the last duplicates.
+    ///
+    /// <https://drafts.csswg.org/cssom/#dom-documentorshadowroot-adoptedstylesheets>
+    fn set_adopted_stylesheet(
+        adopted_stylesheets: &mut Vec<Dom<CSSStyleSheet>>,
+        incoming_stylesheets: &[Dom<CSSStyleSheet>],
+        owner: &StyleSheetListOwner,
+    ) -> ErrorResult {
+        if !pref!(dom_adoptedstylesheet_enabled) {
+            return Ok(());
+        }
+
+        let owner_doc = match owner {
+            StyleSheetListOwner::Document(doc) => doc,
+            StyleSheetListOwner::ShadowRoot(root) => root.owner_doc(),
+        };
+
+        for sheet in incoming_stylesheets.iter() {
+            // > If value’s constructed flag is not set, or its constructor document is not equal
+            // > to this DocumentOrShadowRoot’s node document, throw a "NotAllowedError" DOMException.
+            if !sheet.constructor_document_matches(owner_doc) {
+                return Err(Error::NotAllowed);
+            }
+        }
+
+        // The set to check for the duplicates when removing the old stylesheets.
+        let mut stylesheet_remove_set = HashSet::with_capacity(adopted_stylesheets.len());
+
+        // Remove the old stylesheets from the StyleSet. This workflow is limited by utilities
+        // Stylo StyleSet given to us.
+        // TODO(stevennovaryo): we could optimize this by maintaining the longest common prefix
+        //                      but we should consider the implementation of ObservableArray as well.
+        for sheet_to_remove in adopted_stylesheets.iter() {
+            // Check for duplicates, only proceed with the removal if the stylesheet is not removed yet.
+            if stylesheet_remove_set.insert(sheet_to_remove) {
+                owner.remove_stylesheet(
+                    StylesheetSource::Constructed(sheet_to_remove.clone()),
+                    sheet_to_remove.style_stylesheet_arc(),
+                );
+                sheet_to_remove.remove_adopter(owner);
+            }
+        }
+
+        // The set to check for the duplicates when adding a new stylesheet.
+        let mut stylesheet_add_set = HashSet::with_capacity(incoming_stylesheets.len());
+
+        // Readd all stylesheet to the StyleSet. This workflow is limited by the utilities
+        // Stylo StyleSet given to us.
+        for sheet in incoming_stylesheets.iter() {
+            // Check for duplicates.
+            if !stylesheet_add_set.insert(sheet) {
+                // The idea is that this case is rare, so we pay the price of removing the
+                // old sheet from the styles and append it later rather than the other way
+                // around.
+                owner.remove_stylesheet(
```

Comment:
I don't think we should deduplicate sheets here. There are cases where:
- The sheet list is [A, B, A]
- We should keep both A. If we keep only one of them, the cascade result is wrong

Example here: https://github.com/web-platform-tests/wpt/blob/master/css/css-cascade/layer-stylesheet-sharing.html

Reply:
I just tried to replicate it with `adoptedStylesheets`. It seems that Blink is keeping the first one while Gecko is keeping the last one, and both are wrong.
- https://jsfiddle.net/d8hxyn5c/1/  Chrome is red, FF is green
- https://jsfiddle.net/d8hxyn5c/2/  Chrome is green, FF is red

So to make things correct, we must keep both.